File: jianzhi_offer/No.11 旋转数组的最小数字/solution.py
class Solution:
    def minArray(self, numbers: List[int]) -> int:
        min,max=0,len(numbers)-1
        while(min<max):
            mid=int((min+max)/2)
            if numbers[mid]>numbers[max]:
                min=mid+1
            elif numbers[mid]<numbers[max]:
                max=mid
            else:
                max-=1
        return numbers[min]

File: jianzhi_offer/No.11 旋转数组的最小数字/solution.cpp

class Solution {
public:
    int minArray(vector<int>& numbers) {
        int min=0;
        int max=numbers.size()-1;
        int mid;
        while(min<max){
            mid=(max+min)/2;
            if (numbers[mid]>numbers[max]){
                min=mid+1;
            }
            else if(numbers[mid]==numbers[max]){
                if (numbers[max-1]==numbers[max])
                    max=max-1;
                else
                    min=mid+1;
            }
            else{
                max=mid;
            }
        }
        return numbers[min];
    }
};